pymeasure/instruments/instrument.py

```diff
@@ -191,103 +227,195 @@ class attributes that are a DynamicProperty. Check also for class variables
+        if name in ("_special_names", "__dict__"):
             return super().__getattribute__(name)
-        if hasattr(self, '_special_names'):
+        if hasattr(self, "_special_names"):
             if name in self._special_names:
                 raise AttributeError(
-                    f"{name} is a reserved variable name and it cannot be read")
+                    f"{name} is a reserved variable name and it cannot be read"
+                )
         return super().__getattribute__(name)
 
     @property
     def complete(self):
-        """ This property allows synchronization between a controller and a device. The Operation Complete
+        """This property allows synchronization between a controller and a device. The Operation Complete
         query places an ASCII character 1 into the device's Output Queue when all pending
         selected device operations have been finished.
         """
         if self.SCPI:
-            return self.ask("*OPC?").strip()
+            ready = self.ask_no_lock("*OPC?").strip()
+            if ready == "1":
+                return True
+            elif ready == "0":
+                return False
+            else:
+                return None
         else:
-            raise NotImplementedError("Non SCPI instruments require implementation in subclasses")
+            raise NotImplementedError(
+                "Non SCPI instruments require implementation in subclasses"
+            )
 
     @property
-    def status(self):
-        """ Requests and returns the status byte and Master Summary Status bit. """
-        if self.SCPI:
-            return self.ask("*STB?").strip()
-        else:
-            raise NotImplementedError("Non SCPI instruments require implementation in subclasses")
+    @abc.abstractmethod
+    def status(self) -> str:
+        """Checks the status of the system. Must be implemented in all concrete drivers."""
 
     @property
     def options(self):
-        """ Requests and returns the device options installed. """
+        """Requests and returns the device options installed."""
         if self.SCPI:
             return self.ask("*OPT?").strip()
         else:
-            raise NotImplementedError("Non SCPI instruments require implementation in subclasses")
+            raise NotImplementedError(
+                "Non SCPI instruments require implementation in subclasses"
+            )
 
-    @property
-    def id(self):
-        """ Requests and returns the identification of the instrument. """
+    def check_errors_and_wait(func):
+        """Decorator used to check errors and wait until the system is ready
+        to process the next command
+        """
+
+        def wait_and_check_errors(self, *args, **kwargs):
+            with self._lock:
+                self._wait_until_ready()
+                result = func(self, *args, **kwargs)
+                self.check_errors()
+                return result
+
+        return wait_and_check_errors
```

Comment:
Don't have to do it, but I think it's convention to call the "internal" method `wrapper`, and it would also save you from defining `wait_and_check_errors` as well as `check_errors_and_wait`.

```suggestion
    def wait_and_check_errorsfunc):
        """
        Decorator used to wait until the system is ready to process the next
        command, and subsequently check for errors.
        """

        def wrapper(self, *args, **kwargs):
            with self._lock:
                self._wait_until_ready()
                result = func(self, *args, **kwargs)
                self.check_errors()
                return result

        return wrapper
```